opencog/query/PatternMatchEngine.cc

```diff
@@ -1077,6 +1077,13 @@ bool PatternMatchEngine::tree_compare(const PatternTermPtr& ptm,
 	if (hp->is_node() and hg->is_node())
 		return node_compare(hp, hg);
 
+	// CHOICE_LINK's are multiple-choice links. As long as we can
+	// can match one of the sub-expressions of the ChoiceLink, then
+	// the ChoiceLink as a whole can be considered to be grounded.
+	//
+	if (CHOICE_LINK == tp)
+		return choice_compare(ptm, hg);
+
```

Comment:
This code-movement means that the `link_match()` callback is no longer called for choice links, and so that callback never gets a chance to accept or reject the potential matching. 

Now, admittedly, the default callback always accepts choice links, and so perhaps this almost does not matter. The other callback, for the recognizer, does not special-case the choice link.  It is possible that the recognizer is buggy in the way that it treats the choicelink.  The recognizer is not as widely used as the pattern matcher, and has fewer test cases.

Reply:
Ahh, OK, I see -- the point is to move this prior to the fuzzy_match(), which throws everything off. That's good.

I might have an easy fix for all this, then ...